File: algorithm/n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.cpp
class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> result;
        this->traversal(root, result);
        return result;
    }
    
    void traversal(Node* root, vector<int>& list) {
        if (root == nullptr) {
            return;
        }
        list.push_back(root->val);
        for (auto& n : root->children) {
            this->traversal(n, list);
        }
    }
};

File: algorithm/n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.go
package n_ary_tree_preorder_traversal

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Children []*Node
 * }
 */

type Node struct {
	Val      int
	Children []*Node
}

func preorder(root *Node) []int {
	result := make([]int, 0)
	dfs(root, &result)
	return result
}

func dfs(root *Node, result *[]int) {
	if root == nil {
		return
	}

	*result = append(*result, root.Val)

	for _, node := range root.Children {
		dfs(node, result)
	}
}
